src/lib/network_pool/transaction_pool.ml

```diff
@@ -1367,135 +1063,172 @@ struct
+                         .to_yojson locally_generated_dropped ) )
+              ] ;
+        (* register locally generated commands *)
+        if is_sender_local then
+          List.iter add_results ~f:(function
+            | Ok (cmd, _dropped) ->
+                if
+                  not
+                    (Set.mem all_dropped_cmd_hashes
+                       (Transaction_hash.User_command_with_valid_signature.hash
+                          cmd ) )
+                then register_locally_generated t cmd
+            | Error _ ->
+                () ) ;
+        (* finalize the update to the pool *)
         t.pool <- pool ;
         Mina_metrics.(
           Gauge.set Transaction_pool.pool_size
             (Float.of_int (Indexed_pool.size pool)) ;
           Counter.inc_one Transaction_pool.transactions_added_to_pool) ;
-        let trust_record =
-          Trust_system.record_envelope_sender t.config.trust_system t.logger
-            sender
+        (* partition the results *)
+        let accepted, rejected =
+          List.partition_map add_results ~f:(function
+            | Ok (cmd, _dropped) ->
+                Either.First cmd
+            | Error (cmd, error) ->
+                Either.Second (cmd, error) )
         in
-        let rec go txs =
-          let open Interruptible.Deferred_let_syntax in
-          match txs with
-          | [] ->
-              Interruptible.Or_error.return ()
-          | (verified, dropped) :: txs ->
-              let tx =
-                Transaction_hash.User_command_with_valid_signature.command
-                  verified
-              in
-              let tx' = Transaction_hash.User_command.of_checked verified in
-              if Indexed_pool.member t.pool tx' then
-                if is_sender_local then (
-                  [%log' info t.logger]
-                    "Rebroadcasting $cmd already present in the pool"
-                    ~metadata:[ ("cmd", User_command.to_yojson tx) ] ;
-                  register_locally_generated t verified ;
-                  go txs )
-                else
-                  let%bind _ =
-                    trust_record (Trust_system.Actions.Sent_old_gossip, None)
-                  in
-                  go txs
-              else (
-                if is_sender_local then register_locally_generated t verified ;
-                if not (List.is_empty dropped) then
-                  [%log' debug t.logger]
-                    "dropped commands due to transaction replacement: $dropped"
-                    ~metadata:[ ("dropped", Cs.to_yojson dropped) ] ;
-                check_dropped (Sequence.of_list dropped) ;
-                go txs )
+        (* determine if we should re-broadcast this diff *)
+        let decision =
+          if
+            List.exists rejected ~f:(fun (_, error) ->
+                Diff_error.grounds_for_diff_rejection error )
+          then `Reject
```

Comment:
Why do we accept some transactions from the diff that is not broadcast-able?

Reply:
Could this cause mem pool inconsistencies?

Reply:
It could contribute to it in theory, but I don't think it would be a major contributor. This basically only happens if the txn pool diff contains a malicious transaction (eg: invalid signature). We would only accept still-valid txns from the diff, but would refuse to send the diff to another peer since it contained some malicious data.

I'm not sure this behavior is the best, but I'm also not certain if it's important to update now. I think the issue it causes is very small and it's unlikely we will see it in practice, so I'm inclined to just wait for a future txn pool rewrite for this edge case.